File: cpp/organism/plant/species/Guarana.cpp
//
// Created by fgrel on 18.04.2025.
//

#include "Guarana.h"

void Guarana::reproduce() {
  Point newPosition = getNewPosition();
  if (newPosition == position)
    return; // Ensure a new position is valid

  world.killPlantsOnPosition(newPosition);
  world.queueOrganismAddition(new Guarana(newPosition, world));
}

double Guarana::getSpreadProbability() const {
  return spreadProbability;
}

bool Guarana::collision(Organism &other) {
  other.setStrength(other.getStrength() + 3);
  world.getLogger().logEvent("Guarana has increased the strength of " +
                              other.getSymbol() + " by 3.");
  die();
  return false;
}

Organism* Guarana::clone() const {
    return new Guarana(*this); // Tworzy kopię obiektu za pomocą konstruktora kopiującego
}

File: java/src/organism/plant/species/Guarana.java
package organism.plant.species;

import constants.Constants;
import logger.EventLogger;
import organism.Organism;
import organism.plant.Plant;
import world.World;

public class Guarana extends Plant {

    public Guarana(int x, int y, World world) {
        super(x, y, Constants.Plant.Guarana.Symbol, Constants.Plant.Guarana.Strength, Constants.Plant.Initiative, world);
    }

    @Override
    protected void reproduce() {
        int[] newPosition = getNewPosition();
        if (newPosition[0] == x && newPosition[1] == y) return;
        world.addOrganism(new Guarana(newPosition[0], newPosition[1], world));
        EventLogger.getInstance().log("Guarana spread to (" + newPosition[0] + ", " + newPosition[1] + ")");
    }

    @Override
    protected double getSpreadProbability() {
        return Constants.Plant.Guarana.SpreadProbability;
    }

    @Override
    public boolean collision(Organism other) {
        other.setStrength(other.getStrength() + 3);
        EventLogger.getInstance().log("Guarana increased the strength of " + other.getClass().getSimpleName() + " by 3 at (" + x + ", " + y + ")");
        die();
        return false;
    }
}